rust/src/csvreader.rs

```diff
@@ -0,0 +1,263 @@
+    ) -> Self {
+        let csv_reader = csv::ReaderBuilder::new()
+            .has_headers(has_headers)
+            .from_reader(BufReader::new(file));
+
+        let record_iter = csv_reader.into_records();
+        CsvReader {
+            schema: schema.clone(),
+            projection,
+            record_iter,
+            batch_size,
+        }
+    }
+}
+
+macro_rules! build_primitive_array {
+    ($ROWS:expr, $COL_INDEX:expr, $TY:ty) => {{
+        let mut builder = PrimitiveArrayBuilder::<$TY>::new($ROWS.len() as i64);
+        for row_index in 0..$ROWS.len() {
+            match $ROWS[row_index].get(*$COL_INDEX) {
+                Some(s) if s.len() > 0 => builder.push(s.parse::<$TY>().unwrap()).unwrap(),
+                _ => builder.push_null().unwrap(),
+            }
+        }
+        Ok(Arc::new(builder.finish()) as ArrayRef)
+    }};
+}
+
+impl CsvReader {
+    /// Read the next batch of rows
+    pub fn next(&mut self) -> Option<Result<Arc<RecordBatch>, ArrowError>> {
+        // read a batch of rows into memory
+        let mut rows: Vec<StringRecord> = Vec::with_capacity(self.batch_size);
+        for _ in 0..self.batch_size {
+            match self.record_iter.next() {
+                Some(Ok(r)) => {
+                    rows.push(r);
+                }
+                Some(Err(_)) => {
+                    return Some(Err(ArrowError::ParseError(
+                        "Error reading CSV file".to_string(),
+                    )));
+                }
+                None => break,
+            }
+        }
+
+        // return early if no data was loaded
+        if rows.len() == 0 {
+            return None;
+        }
+
+        let projection: Vec<usize> = match self.projection {
+            Some(ref v) => v.clone(),
+            None => self
+                .schema
+                .fields()
+                .iter()
+                .enumerate()
+                .map(|(i, _)| i)
+                .collect(),
+        };
+
+        let arrays: Result<Vec<ArrayRef>, ArrowError> = projection
+            .iter()
+            .map(|i| {
+                let field = self.schema.field(*i);
+
+                match field.data_type() {
+                    &DataType::Boolean => build_primitive_array!(rows, i, bool),
+                    &DataType::Int8 => build_primitive_array!(rows, i, i8),
+                    &DataType::Int16 => build_primitive_array!(rows, i, i16),
+                    &DataType::Int32 => build_primitive_array!(rows, i, i32),
+                    &DataType::Int64 => build_primitive_array!(rows, i, i64),
+                    &DataType::UInt8 => build_primitive_array!(rows, i, u8),
+                    &DataType::UInt16 => build_primitive_array!(rows, i, u16),
+                    &DataType::UInt32 => build_primitive_array!(rows, i, u32),
+                    &DataType::UInt64 => build_primitive_array!(rows, i, u64),
+                    &DataType::Float16 => build_primitive_array!(rows, i, f32),
```

Comment:
This doesn't look right - half float is stored with 2 bytes but here we are making them `f32` (4 bytes). Maybe we should leave it as a TODO for now.

Reply:
I removed support for `Float16` so this would now fail with an unsupported data type error, which seems sensible